# Task 2: Enhanced House Price Prediction — Model Comparison
**Dataset:** California Housing | **Goal:** Compare multiple regression models with proper preprocessing

## Step 1: Import Required Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import joblib
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score

## Step 2: Load the Dataset

In [2]:
data = fetch_california_housing(as_frame=True)
df = pd.concat([data.data, data.target.rename("HousePrice")], axis=1)
print("Dataset shape:", df.shape)
df.head()

Dataset shape: (20640, 9)


MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,HousePrice
8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521


## Step 3: Separate Features and Target Variable

In [3]:
X = df.drop("HousePrice", axis=1)
y = df["HousePrice"]

## Step 4: Feature Scaling
Standardize features so no single feature dominates due to scale differences.

In [4]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

## Step 5: Train-Test Split

In [5]:
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42
)

## Step 6: Train Multiple Models

In [6]:
models = {
    "Linear Regression": LinearRegression(),
    "Ridge Regression": Ridge(alpha=1.0),
    "Decision Tree": DecisionTreeRegressor(max_depth=5)
}

## Step 7: Model Evaluation and Comparison

In [7]:
results = {}
for name, model in models.items():
    model.fit(X_train, y_train)
    predictions = model.predict(X_test)
    rmse = np.sqrt(mean_squared_error(y_test, predictions))
    r2 = r2_score(y_test, predictions)
    results[name] = {"RMSE": round(rmse, 4), "R2 Score": round(r2, 4)}

results_df = pd.DataFrame(results).T
print("Model Performance Comparison:")
print(results_df.to_string())

Model Performance Comparison:
                     RMSE  R2 Score
Linear Regression  0.7456    0.5758
Ridge Regression   0.7456    0.5758
Decision Tree      0.7242    0.5997


**Interpretation:**
- Lower RMSE → better prediction accuracy
- Higher R² Score → better explanatory power
- **Decision Tree (max_depth=5)** wins with RMSE=0.7242 and R²=0.5997

## Step 8: Visual Performance Validation

In [8]:
best_name = results_df["R2 Score"].astype(float).idxmax()
best_model = models[best_name]
y_pred = best_model.predict(X_test)

plt.figure(figsize=(8, 6))
plt.scatter(y_test, y_pred, alpha=0.4, label="Predictions")
plt.plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], color="red", label="Ideal")
plt.xlabel("Actual House Prices")
plt.ylabel("Predicted House Prices")
plt.title(f"Actual vs Predicted — {best_name}")
plt.legend()
plt.tight_layout()
plt.savefig("actual_vs_predicted.png", dpi=150)
plt.show()
print(f"Best Model: {best_name}")
print("Chart saved: actual_vs_predicted.png")

Best Model: Decision Tree
Chart saved: actual_vs_predicted.png


## Optional: Save the Best Model

In [9]:
joblib.dump(best_model, "best_model.pkl")
print(f"Saved: {best_name} → best_model.pkl")

Saved: Decision Tree → best_model.pkl


## Summary

| Step | Action | Detail |
|------|--------|--------|
| Data Loading | California Housing Dataset | 20,640 rows × 9 columns |
| Preprocessing | StandardScaler | All 8 features normalized |
| Train/Test Split | 80/20 split | random_state=42 |
| Models Trained | Linear Regression, Ridge Regression, Decision Tree | 3 algorithms |
| Evaluation Metrics | RMSE + R² Score | Per model |
| Best Model | **Decision Tree (max_depth=5)** | RMSE=0.7242, R²=0.5997 |
| Output Files | actual_vs_predicted.png, best_model.pkl | Saved to Task2/ |